src/Graph/Trees/BallTree.php

```diff
@@ -150,9 +150,14 @@ public function grow(Labeled $dataset) : void
             if ($right->numRows() > $this->maxLeafSize) {
                 $node = Ball::split($right, $this->kernel);
 
-                $current->attachRight($node);
+                [$nleft, $nright] = $node->groups();
 
-                $stack[] = $node;
+                if ($left->empty() && $nleft->empty() && $right == $nright) {
```

Comment:
Would ...

```php
if ($node->radius() === 0.0) {
```

work here instead?

Reply:
Yes, I think your logic about hypersphere radius is correct, so this seems a nice improvement of the check.